### PR TITLE
Show comment API errors with `OfflineWithFeedback`

### PR DESCRIPTION
**Pull request by marcaaron**
### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/218572

### Tests
1. Test together with https://github.com/Expensify/Web-Expensify/pull/34538
1. Trigger an error by throwing an ExpError somewhere in the try/catch [here](https://github.com/Expensify/Web-Expensify/blob/c09ab4be6890efccafff09937679910ac313eef8/lib/ReportAPI.php#L337)
1. Verify that the error message appears below the chat message
1. Verify that the sidebar has a red dot (red brick road indicator)
1. Dismiss the error
2. Verify the red brick road indicator is removed.

- [ ] Verify that no errors appear in the JS console

#### Contributor (PR Author) Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] iOS / native
    - [x] Android / native
    - [x] iOS / Safari
    - [x] Android / Chrome
    - [x] MacOS / Chrome
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] Any functional components have the `displayName` property
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

<details>
<summary><h4>PR Reviewer Checklist</h4>

The Contributor+ will copy/paste it into a new comment and complete it after the author checklist is completed
</summary>

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] iOS / native
    - [ ] Android / native
    - [ ] iOS / Safari
    - [ ] Android / Chrome
    - [ ] MacOS / Chrome
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product was added in all `src/languages/*` files
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by tagging the marketing team on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] Any functional components have the `displayName` property
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

</details>

### QA Steps

- [ ] Verify that no errors appear in the JS console

### CODE REVIEW

**Comment by marcaaron** [comment]
Style on this is looking a bit rough but the opacity and errors are showing up...

Screen Shot 2022-08-08 at 3 52 45 PM

**To Do:**

- [ ] Fix the styles
- [ ] Work on getting the red brick road indicators to show up in the sidebar

File: src/libs/actions/Report.js
Comment (marcaaron) [inline src/libs/actions/Report.js:419]: This is kind of an MVP solution that I'm removing now. If we leave it in each time the report actions are fetched we will remove any "optimistic actions". That includes any pending actions with errors which breaks the new design. It was always the plan to remove this code so if we end up with a problem related to "stuck actions" then we'll just need to approach it another way. Though, I'm not sure if "stuck comments" are really a thing anymore.

**Comment by marcaaron** [comment]
I think this is almost ready.  This is what the errors look like atm:

Screen Shot 2022-08-09 at 11 13 44 AM
Screen Shot 2022-08-09 at 11 13 37 AM
Screen Shot 2022-08-09 at 11 12 40 AM

cc @shawnborton to see if there are any notes!

**Comment by shawnborton** [comment]
Looks pretty good! Make sure the right side of the error block also has 20px padding, so that the right icon aligns well with the content above and below it. Also, is this nested within the message above it? As in, if you hover over the message row with the error, does the row highlight color show behind the error as well?

File: src/pages/home/report/ReportActionsList.js
Comment (neil-marcellini) [inline src/pages/home/report/ReportActionsList.js:145]: What if `pendingAction` or `errors` doesn't exist on the `action`. I think `lodashGet` should be used here.

**Review by neil-marcellini — CHANGES_REQUESTED** [review]
Looks good overall and it works well. I have some questions and a couple things that I think should be changed. As Shawn said the error needs some right margin.

The error row is separate from the comment when highlighted. Personally I think it looks fine.
image

Also, the tooltip doesn't look quite right
image

File: src/libs/OptionsListUtils.js
Comment (marcaaron) [reply src/libs/OptionsListUtils.js:202]: Mostly because I have a preference for functions that return a consistent type.

File: src/libs/OptionsListUtils.js
Comment (marcaaron) [reply src/libs/OptionsListUtils.js:199]: The report might not exist at all since we could have a personal details option.

File: src/libs/OptionsListUtils.js
Comment (marcaaron) [reply src/libs/OptionsListUtils.js:603]: No, but I don't have any specific concerns with passing the `reportActions` here. We could create a more specific `createReportOption()` alternatively, but unsure whether it's necessary.

File: src/libs/actions/Report.js
Comment (marcaaron) [reply src/libs/actions/Report.js:64]: We were:

- adding all optimistic actions `clientID` to an array on each report object
- removing them later whichever one happens first
    - The real action created in the server shows up in Pusher
    - The latest actions for the report are fetched

Clearing out the actions was done to take care of an edge case where:

- A report comment would fail to be created in the server
- No error message would be displayed and there would be no way to dismiss the message
- These comments would get "stuck" in a kind of limbo of never being created and unable to be removed 
- So to counteract that we would clear these out when fetching new actions

Ultimately, I think this is not as much of a problem any more since we are handling the failure cases and giving the user a way to dismiss a message that could not be created for some reason.

File: src/pages/home/report/ReportActionsList.js
Comment (marcaaron) [reply src/pages/home/report/ReportActionsList.js:145]: if `pendingAction` or `errors` doesn't exist, but `item.action` does exist then they will just be `undefined` and we are probably OK because of this:

https://github.com/Expensify/App/blob/f5f076a13bb4a991dbeaf32a8f7a7b0625e8a3cc/src/components/OfflineWithFeedback.js#L49-L51

If the action itself is `undefined` then the JS would throw a `TypeError`, but it doesn't seem possible to me.

File: src/libs/OptionsListUtils.js
Comment (neil-marcellini) [reply src/libs/OptionsListUtils.js:202]: Makes sense 👍

File: src/libs/OptionsListUtils.js
Comment (neil-marcellini) [reply src/libs/OptionsListUtils.js:199]: Ah ok, got it.

File: src/libs/actions/Report.js
Comment (neil-marcellini) [reply src/libs/actions/Report.js:64]: Makes sense. Thanks for the explanation.

**Comment by marcaaron** [comment]
Yeah so I'm not too sure what exactly is happening with the tooltip tbh. Tried switching from margin to padding to see if that helped. 

Here's what this looks like with improved spacing on the right side and everything under the "hover"

Screen Shot 2022-08-09 at 1 57 10 PM
Screen Shot 2022-08-09 at 1 56 53 PM
Screen Shot 2022-08-09 at 1 57 20 PM

Unsure if this is a blocker or something we might improve later, but here's some feedback...

1. It's not entirely clear that the "Close" button will remove the message. Should it say something like "Delete Comment" instead? Do we care?

2. The `x` feels a bit far from the error message on desktop and having an `x` in the bottom right corner sort of betrays my expectations (x usually in the top right or left to close or dismiss something).

**Comment by marcaaron** [comment]
This effect is not as noticeable on mobile sizes ->
 
Screen Shot 2022-08-09 at 2 04 59 PM

**Review by neil-marcellini — DISMISSED** [review]
> It's not entirely clear that the "Close" button will remove the message. Should it say something like "Delete Comment" instead? Do we care?

I think it's fine. If this command fails I suspect the error message will give some context that will make it clear that the message was not sent / saved.

> The x feels a bit far from the error message on desktop and having an x in the bottom right corner sort of betrays my expectations (x usually in the top right or left to close or dismiss something).

Yeah, it doesn't look super great, but that's only the case for very large screen sizes. I think it looks pretty good on a laptop sized screen or smaller. The hover actions like "copy to clipboard" are also pushed to the far right side and we seem to be okay with that.

Also, the tooltip looks different now and I think it's satisfactory at this point. We can always make small tweaks in the future.
image

I'll approve but I'm curious to hear what @shawnborton thinks before we merge.

File: src/components/OptionRow.js
Comment (PauloGasparSv) [inline src/components/OptionRow.js:210]: Is there a reason for the double brackets `[[ ]]` on this style?

**Review by PauloGasparSv — DISMISSED** [review]

**Comment by shawnborton** [comment]
Both the tooltip and the "x" being far off to the right are totally cool with me, so I'm good with merging this.

**Review by shawnborton — DISMISSED** [review]

File: src/components/OptionRow.js
Comment (marcaaron) [reply src/components/OptionRow.js:210]: Oh whoa weird I copied these styles from somewhere else, but the double brackets are definitely not necessary (and surprised it even worked).

**Review by PauloGasparSv — APPROVED** [review]

**Review by neil-marcellini — APPROVED** [review]

**Comment by marcaaron** [comment]
I'm gonna merge this one to unblock @neil-marcellini and @PauloGasparSv. I tested against the staging Web-Expensify and don't think this needs to be on HOLD.

**Comment by marcaaron** [comment]
Tests passed

**Pull request merged**
